lib/include/openamp/rpmsg_rpc_client_server.h

```diff
@@ -0,0 +1,246 @@
+	int (*cb_function)(void *data, struct rpmsg_rpc_svr *rpcs);
+};
+
+/**
+ * struct rpmsg_rpc_request - rpc request message
+ *
+ * @id: service id
+ * @params: request params
+ *
+ */
+struct rpmsg_rpc_request {
+	uint32_t id;
+	unsigned char params[MAX_BUF_LEN];
+};
+
+/**
+ * struct rpmsg_rpc_answer - rpc request message
+ *
+ * @id: service id
+ * @status: status of rpc
+ * @params: answer params
+ *
+ */
+struct rpmsg_rpc_answer {
+	uint32_t id;
+	int32_t status;
+	unsigned char params[MAX_BUF_LEN];
+};
+
+/**
+ * struct rpmsg_rpc_services - table for services
+ *
+ * @id: service id
+ * @respcb: id callback
+ *
+ */
+struct rpmsg_rpc_services {
+	uint32_t id;
+	struct rpmsg_rpc_syscall_cb respcb;
+};
+
+/**
+ * struct rpmsg_rpc_client_services - table for client services
+ *
+ * @id: service id
+ * @app_cb: id callback
+ *
+ */
+struct rpmsg_rpc_client_services {
+	uint32_t id;
+	app_cb cb;
+};
+
+/**
+ * struct rpmsg_rpc_svr - server remote procedure call data
+ *
+ * RPMsg RPC will send request to endpoint
+ *
+ * @ept: rpmsg_endpoint structure
+ * @services: service table
+ * @n_services: number of services
+ *
+ */
+struct rpmsg_rpc_svr {
+	struct rpmsg_endpoint ept;
+	const struct rpmsg_rpc_services *services;
+	unsigned int n_services;
+};
+
+/**
+ * struct rpmsg_rpc_data - remote procedure call for global data
+ *
+ * RPMsg RPC will send request to remote and
+ * wait for callback.
+ *
+ * @ept: rpmsg_endpoint structure
+ * @shutdown_cb: shutdown callback function
+ * @services: service table
+ * @n_services: number of services
+ *
+ */
+struct rpmsg_rpc_data {
+	struct rpmsg_endpoint ept;
+	rpmsg_rpc_shutdown_cb shutdown_cb;
+	const struct rpmsg_rpc_client_services *services;
+	unsigned int n_services;
+};
+
+/**
+ * rpmsg_rpc_client_release - release RPMsg remote procedure call
+ *
+ * This function is to release remoteproc procedure call
+ * global data.
+ *
+ * @rpc: pointer to the global remote procedure call data
+ *
+ */
+void rpmsg_rpc_client_release(struct rpmsg_rpc_data *rpc);
+
+/**
+ * rpmsg_rpc_client_init - initialize RPMsg remote procedure call
+ *
+ * This function is to initialize the remote procedure call
+ * global data. RPMsg RPC will send request to remote and
+ * wait for callback and load services to table
+ *
+ * @rpc: pointer to the global remote procedure call data
+ * @rdev: pointer to the rpmsg device
+ * @ept_name: name of the endpoint used by RPC
+ * @ept_addr: address of the endpoint used by RPC
+ * @ept_raddr: remote address of the endpoint used by RPC
+ * @shutdown_cb: shutdown callback function
+ * @services: pointer to service table
+ * @len: length of table
+ *
+ * return 0 for success, and negative value for failure
+ */
+int rpmsg_rpc_client_init(struct rpmsg_rpc_data *rpc,
+			  struct rpmsg_device *rdev,
+			  const char *ept_name, uint32_t ept_addr,
+			  uint32_t ept_raddr,
+			  rpmsg_rpc_shutdown_cb shutdown_cb,
+			  const struct rpmsg_rpc_client_services *services,
+			  int len);
+
```

Comment:
Sorry just notice that now...
The name service and endpoint address should not be part of the rpmsg_rpc_client_init and rpmsg_rpc_server_init param. a sepecific APi will be added if someone need to specify them.
```
int rpmsg_rpc_client_init(struct rpmsg_rpc_data *rpc,
			  struct rpmsg_device *rdev,
			  rpmsg_rpc_shutdown_cb shutdown_cb,
			  const struct rpmsg_rpc_client_services *services,
			  int len);
```
i describe impact in the function body below